### ISSUE TITLE
Repeatedly building with `dune build` takes upwards of 10 minutes every time.

### ISSUE THREAD

**Issue by kindaro** [coq/coq#13332]
#### Description of the problem

Every time I run `dune build`, it takes upwards of **10 minutes** to complete. For most of the duration of the build, it displays the following line:

    Done: 14010/14020 (jobs: 1)

I infer that there is a specific task that takes all the time.

A brief look at `ps -o command` tells me that the task is this:

    make -j 2 BIN= COQLIB=/srv/src/coq/_build/install/default/lib/coq PRINT_LOGS=1 UNIT_TESTS=unit-tests

— That is, `dune` is running the test suite, 2 sub-tasks at a time. This is wrong in 2 ways:
* I have 8 cores. Tests can be executed in parallel. Therefore, the task could execute 4 times faster.
* I asked to build, not check. There is a separate command `dune runtest` that I would use if I consciously wanted to run the full test suite. Ordinarily I would not like to throw away 10 minutes and would rather run individual tests that pertain to the work I am doing. _(See #13328 for whether this is possible with `dune`.)_

Is there a good reason for this behaviour to be default?

The promise of a build system is that only as much work is performed as dictated by changes in the source code. So, repeated invocations of a build system should only take as much time as it takes to see that no files have been changed — under a minute for a project the size of Coq. Note that the default invocation of `make` respects this and takes about **one second**.

I realize there is a variety of targets and other options I can give to `dune`. It is not trivial to figure out what the correct invocation is, or even if such one exists! That it takes hours of fiddling to figure out how to build a project properly means that it is very difficult for anyone but a seasoned frequent contributor to do meaningful work. I refer to #13328 again for a longer explanation of how sad I am and why.

#### Proposition

I propose that `dune` be instructed not to run any tests with the default invocation, and also that, when instructed specifically to run tests, it utilizes `nproc` − 1 cores. _(One free core for browsing the Coq issue tracker in the mean time!)_

#### `dune` Version

    2.5.0

**Comment by Zimmi48** [comment]
>Is there a good reason for this behaviour to be default?

I agree this is not a sane behavior and it should be relatively high priority to change this. We can tell people to not use the default target, but it would be easier to just change the default target (since it can be done).

**Comment by ejgallego** [comment]
This is due to `dune build` being an alias for `dune build @all`, whether we want `@all` to include tests , I don't know, I'd say yes.

So the issue here is that the test-suite has not been ported to dune yet, once it is, `dune build` will correctly understand what tests need to be re-run and this issue will go away.

**Comment by ejgallego** [comment]
Actually the right targets are documented in `Makefile.dune`

**Comment by Zimmi48** [comment]
> This is due to `dune build` being an alias for `dune build @all`, whether we want `@all` to include tests , I don't know, I'd say yes.

Is this common in the OCaml world to include tests in `@all`?

Besides, `dune build` is actually an alias for `dune build @@default`. We are free to define what the `default` alias mean but the default is `@all`. See https://dune.readthedocs.io/en/stable/usage.html#default-alias.

I would favor defining a smaller default (people can still do `dune build @all` if they need), regardless of whether we wish to include the test-suite in `@all` or not.

**Comment by kindaro** [comment]
@ejgallego  This sounds dismissive. As I already said, there are very many targets and options, and I have no idea what I should be building. I only want to make a small contribution. It is not rational for every prospective contributor to study the build scripts in detail and try out every little flag. It is not instant. It takes days. _«Build everything»_ would be fine, but no such target appears to me. And no, it should not include _running_ the test suite.

Building tests with `dune` has been in works for at least as long as I have been around — about half a year — and I see no progress. In other words, things are broken for half a year already. So your saying _«once the test-suite has been ported to dune, this issue will go away»_ does not make me any less sad.

**Comment by ejgallego** [comment]
> @ejgallego  This sounds dismissive. As I already said, there are very many targets and options, and I have no idea what I should be building. I only want to make a small contribution. It is not rational for every prospective contributor to study the build scripts in detail and try out every little flag. It is not instant. It takes days. _«Build everything»_ would be fine, but no such target appears to me. And no, it should not include _running_ the test suite.

Sorry but I don't see how any of this is dismissive, in fact I acknowledge all your reports. See below.

> Building tests with `dune` has been in works for at least as long as I have been around — about half a year — and I see no progress. In other words, things are broken for half a year already. So your saying _«once the test-suite has been ported to dune, this issue will go away»_ does not make me any less sad.

We are sticking to purely technical discussion here, so at least us we are neither sad or happy. There is a mix of a few things here:

- dune support is still experimental, a bit of "use at your own risk". This is of course not ideal but we have been doing all we can.
- a few of the issues, for example what targets to build, are related to documentation. We do indeed really appreciate improvements to documentation, and indeed we thought that using the recommended entry point `make -f Makefile.dune` was already providing a good enough guide
- in regards to this bug report, it is really a duplicate of "test suite should be ported to dune", then this bug would not exist, so it is not dismissive, but indeed from a technical point of view that is really the culprit. Unfortunately doing such port requires a lot of time so didn't manage yet :/

So what I think a reasonable course of action here is:
- improve docs, make a bit more clear what works and what doesn't
- be clear about what features we expect to work, and which we don't. For example, we don't aim to support a hybrid make / dune OCaml build. We do aim to support a smooth test-suite integration. I'm sure you will find many more so please keep providing feedback.

**Comment by kindaro** [comment]
> Sorry but I don't see how any of this is dismissive, in fact I acknowledge all your reports. … We do indeed really appreciate improvements to documentation, and indeed we thought that using the recommended entry point `make -f Makefile.dune` was already providing a good enough guide

You are saying previously:

> Actually the right targets are documented in `Makefile.dune`

No, they are not. If they were, there would be no issue. See my anticipation of this message in the opening message:

> I realize there is a variety of targets and other options I can give to dune. It is not trivial to figure out what the correct invocation is, or even if such one exists!

I consulted all available documentation before opening this issue. I made a reasonable effort. It is not feasible to infer the correct `dune` invocation from the documentation alone, because it does not explain anything. Do I need `coqtop`, or `coq`, or `states`, or `check`? How can I decide? I suppose the assorted notes that there are suffice as a reminder for a frequent contributor, but they are not suitable for a person to start being minimally productive. A reasonable default should be advised, and then why not even set it as an actual _default_ of the invocation?

So, at present, what build targets and options should I give to `dune` for optimal result? I would like to build everything as it does now, but without running the test suite. Can I say something like `dune build --no-test-suite`?

I understand your point that the current default is the right one to have and so should be kept. But is it reasonable? If you have a broken system for such extent of time as half a year, you should make a _temporary_ default that works _now_. You can set it back to the _«correct»_ value when the `dune` engine can actually run those tests. The small effort to flip a switch versus hours lost waiting for unnecessary build tasks.

**Comment by Zimmi48** [comment]
> A reasonable default should be advised, and then why not even set it as an actual _default_ of the invocation?

Agreed, and this is the main point of the issue (making it distinctive from a "port test-suite to Dune" one).

**Comment by kindaro** [comment]
By the way, what is the difference between `dune build` and `dune runtest` given that `dune build` runs the test suite?

**Comment by ejgallego** [comment]
> > A reasonable default should be advised, and then why not even set it as an actual _default_ of the invocation?
> 
> Agreed, and this is the main point of the issue (making it distinctive from a "port test-suite to Dune" one).

Well, to me what we have seems OK:
```
Welcome to Coq's Dune-based build system. Common developer targets are:

  - states: build a minimal functional coqtop
  - world:  build all public binaries and libraries
  - watch:  build all public binaries and libraries [continuous build]
  - check:  build all ML files as fast as possible
  - test-suite: run Coq's test suite
```
feel free to submit improvement. Note that we don't suggest calling dune directly yet, unless you are an expert.

**Comment by kindaro** [comment]
I shared my negative experience, explained a real problem and put forward a question:

> … Do I need `coqtop`, or `coq`, or `states`, or `check`? How can I decide? …
> 
> So, at present, what build targets and options should I give to `dune` for optimal result? I would like to build everything as it does now, but without running the test suite. Can I say something like `dune build --no-test-suite`?

If your answer is to go and find out for myself which of the four options does what and which one is the most suitable for me, I can only say that you have not helped at all. If this is the final word of the development team, then I suppose the ticket may be closed.

**Comment by ejgallego** [comment]
@kindaro sorry I don't follow you. The ticket cannot be closed, in fact it contains at least 3 separate bugs:

- documentation on targets seems insufficient [for example `make -f Makefile.dune test-suite` will use `--no-buffer` so you can see progress
- test-suite is not ported to dune, and it is thus rebuilt even if nothing changed [note that Make has the same problem by the way]
- the default dune target may be inappropiate, tho when the previous point is solved it won't matter so much [ `@check` is the standard ML regular one for example]

**Comment by SkySkimmer** [comment]
Dune default target changed in #14559
